File: src/test/java/me/paulo/casaes/bbop/model/PlayerTest.java
package me.paulo.casaes.bbop.model;

import me.paulo.casaes.bbop.dto.BoltFiredEventDto;
import me.paulo.casaes.bbop.dto.BoltMovedEventDto;
import me.paulo.casaes.bbop.dto.CommandType;
import me.paulo.casaes.bbop.dto.Dto;
import me.paulo.casaes.bbop.dto.EventDto;
import me.paulo.casaes.bbop.dto.EventType;
import me.paulo.casaes.bbop.dto.PlayerDto;
import me.paulo.casaes.bbop.dto.PlayerJoinedEventDto;
import me.paulo.casaes.bbop.dto.PlayerLeftEventDto;
import me.paulo.casaes.bbop.dto.PlayerMovedEventDto;
import me.paulo.casaes.bbop.dto.PlayersListCommandDto;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNotSame;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class PlayerTest {

    @Mock
    private Clock clock;

    @Mock
    private ScoreBoard scoreBoard;

    @Mock
    private Game game;

    private Bolts bolts;

    private Players players;

    @BeforeEach
    public void setup() {
        MockitoAnnotations.initMocks(this);
        this.bolts = Bolts.create();
        this.players = Players.create(bolts, clock, scoreBoard);

        doReturn(clock)
                .when(game)
                .getClock();

        doReturn(scoreBoard)
                .when(game)
                .getScoreBoard();

        doReturn(players)
                .when(game)
                .getPlayers();

        doReturn(bolts)
                .when(game)
                .getBolts();

        Topics.setGame(game);

        GameEvents.getClientEvents().setConsumer(null);
        GameEvents.getDomainEvents().setConsumer(domainEvent -> {
            Topics.valueOf(domainEvent.getTopic()).consume(domainEvent);
        });


        when(clock.getTime()).thenReturn(0L);

        Config.get().setEnv(Config.Environment.DEV.name());
        Config.get().setPlayerMaxMove(1f);
        Config.get().setMinMove(0.000000001f);
        Config.get().setPlayerMaxAngleDivisor(0.5f);
    }

    @Test
    void testCreate() {
        UUID one = UUID.randomUUID();
        UUID two = UUID.randomUUID();
        Player player = this.players.createOrGet(one);

        assertTrue(this.players.stream()
                .anyMatch(p -> p == player));

        assertTrue(player.is(one));
        assertFalse(player.is(two));

        assertSame(player, this.players.get(one).orElse(null));
        assertEquals(player, this.players.get(one.toString()).orElse(null));

        assertNotSame(player, this.players.get(two).orElse(null));
        assertNotEquals(player, this.players.get(two.toString()).orElse(null));

    }

    @Test
    void testJoin() {
        AtomicReference<Dto> eventReference = new AtomicReference<>(null);
        GameEvents.getClientEvents().setConsumer(eventReference::set);

        UUID one = UUID.randomUUID();
        this.players.createOrGet(one).join();
        PlayerJoinedEventDto event = (PlayerJoinedEventDto) eventReference.get();

        assertNotNull(event);
        assertEquals(EventType.PLAYER_JOINED, event.getEvent());
        assertEquals(one.toString(), event.getPlayerId());
        assertEquals(0f, event.getX());
        assertEquals(0f, event.getY());
        assertEquals(0f, event.getY());
    }

    @Test
    void testJoined() {
        AtomicReference<Dto> eventReference = new AtomicReference<>(null);
        GameEvents.getClientEvents().setConsumer(eventReference::set);

        UUID one = UUID.randomUUID();

        this.players.joined(
                PlayerJoinedEventDto.of(one.toString(), 5, 0f, 1f, 2f)
        );
        PlayerJoinedEventDto event = (PlayerJoinedEventDto) eventReference.get();

        assertNotNull(event);
        assertEquals(EventType.PLAYER_JOINED, event.getEvent());
        assertEquals(one.toString(), event.getPlayerId());
        assertEquals(0f, event.getX());
        assertEquals(1f, event.getY());
        assertEquals(2f, event.getAngle());
        assertEquals(5, event.getShip());
    }

    @Test
    void testLeave() {
        List<Dto> dtos = new ArrayList<>();
        GameEvents.getClientEvents().setConsumer(dtos::add);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.leave();

        List<EventDto> events = dtos
                .stream()
                .filter(dto -> dto.getDtoType() == EventDto.DtoType.EVENT_DTO)
                .map(dto -> (EventDto) dto)
                .collect(Collectors.toList());

        assertEquals(1, events.size());

        EventDto event = events.get(0);
        assertEquals(EventType.PLAYER_LEFT, event.getEvent());
        assertEquals(one.toString(), ((PlayerLeftEventDto) event).getPlayerId());

        assertFalse(this.players.stream()
                .anyMatch(p -> p == player));

    }

    @Test
    void testRequestListOfPlayers() {
        UUID one = UUID.randomUUID();
        UUID two = UUID.randomUUID();
        this.players.createOrGet(one).join();
        this.players.createOrGet(two).join();

        PlayersListCommandDto command = this.players.requestListOfPlayers();

        assertNotNull(command);
        assertEquals(CommandType.LIST_PLAYERS, command.getCommand());
        assertEquals(2, command.getPlayers().size());

        Set<UUID> playerIds = command.getPlayers()
                .stream()
                .map(PlayerDto::getPlayerId)
                .map(UUID::fromString)
                .collect(Collectors.toCollection(HashSet::new));

        assertTrue(playerIds.contains(one));
        assertTrue(playerIds.contains(two));
    }

    @Test
    void testMove() {
        AtomicReference<Dto> eventReference = new AtomicReference<>(null);
        GameEvents.getClientEvents().setConsumer(eventReference::set);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.1f, 0.2f, (float) Math.PI, (float) Math.PI);

        assertEquals(EventDto.DtoType.EVENT_DTO, eventReference.get().getDtoType());
        assertEquals(EventType.PLAYER_MOVED, ((EventDto) eventReference.get()).getEvent());
        PlayerMovedEventDto event = (PlayerMovedEventDto) eventReference.get();

        assertNotNull(event);
        assertEquals(0.1f, event.getX());
        assertEquals(0.2f, event.getY());
        assertEquals((float) Math.PI, event.getAngle());
        assertEquals((float) Math.PI, event.getThrustAngle());

    }

    @Test
    void testBoundedMove() {
        AtomicReference<Dto> eventReference = new AtomicReference<>(null);
        GameEvents.getClientEvents().setConsumer(eventReference::set);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(-1f, 2f, null, null);

        assertEquals(EventDto.DtoType.EVENT_DTO, eventReference.get().getDtoType());
        assertEquals(EventType.PLAYER_MOVED, ((EventDto) eventReference.get()).getEvent());
        PlayerMovedEventDto event = (PlayerMovedEventDto) eventReference.get();

        assertNotNull(event);
        assertEquals(0f, event.getX());
        assertEquals(1f, event.getY());
        assertEquals(0f, event.getAngle());

    }

    @Test
    void testOnlyAngleMove() {
        AtomicReference<Dto> eventReference = new AtomicReference<>(null);
        GameEvents.getClientEvents().setConsumer(eventReference::set);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0f, 0f, (float) Math.PI, (float) Math.PI);

        assertEquals(EventDto.DtoType.EVENT_DTO, eventReference.get().getDtoType());
        assertEquals(EventType.PLAYER_MOVED, ((EventDto) eventReference.get()).getEvent());
        PlayerMovedEventDto event = (PlayerMovedEventDto) eventReference.get();

        assertNotNull(event);
        assertEquals(0f, event.getX());
        assertEquals(0f, event.getY());
        assertEquals((float) Math.PI, event.getAngle());

    }

    @Test
    void testNoMove() {
        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();

        AtomicReference<Dto> eventReference = new AtomicReference<>(null);
        GameEvents.getClientEvents().setConsumer(eventReference::set);

        player.move(0f, 0f, null, null);

        assertNull(eventReference.get());
    }

    @Test
    void testMaxFire() {
        Config.get().setMaxBolts(2);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();

        player.fire();
        player.fire();
        player.fire();


        assertEquals(2, player.getActiveBoltCount());
    }

    @Test
    void testBoltExhaustion() {
        Config.get().setMaxBolts(2);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();

        player.fire();

        assertEquals(1, player.getActiveBoltCount());

        player.boltExhausted();

        assertEquals(0, player.getActiveBoltCount());
    }

    @Test
    void testFireDirection() {
        Config.get().setMaxBolts(2);

        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.5f, 0.5f, (float) Math.PI, (float) Math.PI);

        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().setConsumer(eventReference::set);

        player.fire();

        DomainEvent event = eventReference.get();
        assertNotNull(event);
        assertEquals(EventType.BOLT_FIRED, event.getEvent().getEvent());

        player.fired((BoltFiredEventDto) event.getEvent());

        assertEquals(1, bolts
                .stream()
                .filter(b -> b.isOwnedBy(one))
                .map(Bolt::generateMovedEvent)
                .map(DomainEvent::getEvent)
                .map(b -> (BoltMovedEventDto) b)
                .filter(b -> b.getAngle() == (float) Math.PI)
                .count());

    }

    @Test
    void testCollisionHitBullsEye() {
        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.5f, 0.5f, null, null);

        assertTrue(player.collision(0.45f, 0.45f, 0.55f, 0.55f, 0.05f));
    }

    @Test
    void testCollisionHitWithinRadius() {
        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.54f, 0.54f, null, null);

        assertTrue(player.collision(0.45f, 0.45f, 0.55f, 0.55f, 0.05f));
    }

    @Test
    void testCollisionNoHitInsideSquare() {
        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.4f, 0.6f, null, null);

        assertFalse(player.collision(0.45f, 0.45f, 0.55f, 0.55f, 0.05f));
    }

    @Test
    void testCollisionNoHitOutsideSquareX() {
        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.2f, 0.2f, null, null);

        assertFalse(player.collision(0.45f, 0.45f, 0.55f, 0.55f, 0.05f));
    }

    @Test
    void testCollisionNoHitOutsideSquareY() {
        UUID one = UUID.randomUUID();
        Player player = this.players.createOrGet(one);
        player.join();
        player.move(0.5f, 0.2f, null, null);

        assertFalse(player.collision(0.45f, 0.45f, 0.55f, 0.55f, 0.05f));
    }

    @Test
    void testDestroyedBy() {
        UUID one = UUID.randomUUID();
        UUID two = UUID.randomUUID();
        Player player1 = this.players.createOrGet(one);
        player1.join();
        Player player2 = this.players.createOrGet(two);
        player2.join();

        player1.move(0.5f, 0.5f, 1f, 1f);

        List<DomainEvent> domainEvents = new ArrayList<>();
        GameEvents.getDomainEvents().setConsumer(domainEvents::add);

        player1.destroy(two);

        List<EventDto> events = domainEvents
                .stream()
                .map(DomainEvent::getEvent)
                .collect(Collectors.toList());


        assertEquals(2, events.size());
        assertEquals(EventType.PLAYER_DESTROYED, events.get(0).getEvent());

        assertEquals(EventType.PLAYER_MOVED, events.get(1).getEvent());

        PlayerMovedEventDto playerMovedEventDto = (PlayerMovedEventDto) events.get(1);
        assertEquals(0f, playerMovedEventDto.getX());
        assertEquals(0f, playerMovedEventDto.getY());
        assertEquals(0f, playerMovedEventDto.getAngle());

        verify(scoreBoard, times(1)).updateScore(two, 1);

    }
}

File: src/test/kotlin/me/pcasaes/hexoids/core/domain/model/PlayerTest.kt
package me.pcasaes.hexoids.core.domain.model

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import me.pcasaes.hexoids.core.domain.config.Config
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory.Companion.factory
import me.pcasaes.hexoids.core.domain.model.Bolts.Companion.create
import me.pcasaes.hexoids.core.domain.model.EntityId.Companion.newId
import me.pcasaes.hexoids.core.domain.model.GameEvents.Companion.getClientEvents
import me.pcasaes.hexoids.core.domain.model.GameEvents.Companion.getDomainEvents
import me.pcasaes.hexoids.core.domain.vector.PositionVector
import me.pcasaes.hexoids.core.domain.vector.PositionVector.Companion.of
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import pcasaes.hexoids.proto.Dto
import pcasaes.hexoids.proto.JoinCommandDto
import pcasaes.hexoids.proto.MoveReason
import pcasaes.hexoids.proto.PlayerJoinedEventDto
import pcasaes.hexoids.proto.PlayerMovedEventDto
import java.util.concurrent.atomic.AtomicReference
import kotlin.math.cos


class PlayerTest {
    private val clock = mockk<Clock>(relaxed = true)

    private val scoreBoard = mockk<ScoreBoard>(relaxed = true)

    private val game = mockk<Game>(relaxed = true)

    private val barriers = mockk<Barriers>(relaxed = true)

    private val physicsQueue = mockk<PhysicsQueueEnqueue>(relaxed = true)

    private lateinit var bolts: Bolts

    private lateinit var players: Players

    @BeforeEach
    fun setup() {

        this.bolts = create()
        this.players = Players.create(bolts, clock, scoreBoard, barriers, physicsQueue, factory())
        Assertions.assertEquals(0, this.players.getTotalNumberOfPlayers())
        Assertions.assertEquals(0, this.players.getNumberOfConnectedPlayers())

        every { game.getClock() } returns clock

        every { game.getScoreBoard() } returns scoreBoard

        every { game.getPlayers() } returns players

        every { game.getBolts() } returns bolts

        every { barriers.iterator() } returns emptyList<Barrier>().iterator()

        every { barriers.spliterator() } returns emptyList<Barrier>().spliterator()

        every { barriers.search(any(), any(), any(), any(), any()) } returns emptyList()

        GameTopic.setGame(game)

        getClientEvents().registerEventDispatcher { }
        getDomainEvents().registerEventDispatcher { domainEvent ->
            GameTopic.valueOf(domainEvent.topic!!).consume(
                domainEvent
            )
        }


        every { clock.getTime() } returns 0L

        Config.setPlayerMaxMove(1f)
        Config.setPlayerNameLength(7)
        Config.setMinMove(0.000000001f)
        Config.setPlayerMaxAngleDivisor(0.5f)
        Config.setBoltInertiaEnabled(false)
        Config.setUpdateFrequencyInMillis(50L)
    }

    @Test
    fun testCreate() {
        val one = newId()
        val two = newId()
        val player = this.players.createOrGet(one)

        Assertions.assertTrue(
            this.players
                .any { p -> p === player })

        Assertions.assertTrue(player.hasId(one))
        Assertions.assertFalse(player.hasId(two))

        Assertions.assertSame(player, this.players.get(one))

        Assertions.assertNotSame(player, this.players.get(two))

        Assertions.assertEquals(1, this.players.getTotalNumberOfPlayers())
        Assertions.assertEquals(0, this.players.getNumberOfConnectedPlayers())
    }

    @Test
    fun testJoin() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        this.players.createOrGet(one).join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        this.players.connected(one)
        val event = eventReference.get()!!.getEvent().getPlayerJoined()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(one.getGuid(), event!!.playerId)
        Assertions.assertEquals("one", event.name)

        Assertions.assertEquals(1, this.players.getTotalNumberOfPlayers())
        Assertions.assertEquals(1, this.players.getNumberOfConnectedPlayers())
    }

    @Test
    fun testJoinNoName() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        this.players.createOrGet(one).join(
            JoinCommandDto
                .newBuilder()
                .build()
        )
        this.players.connected(one)
        val event = eventReference.get()!!.getEvent().getPlayerJoined()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(one.getGuid(), event!!.playerId)
        Assertions.assertEquals(one.toString().substring(0, Config.getPlayerNameLength()), event.name)

        Assertions.assertEquals(1, this.players.getTotalNumberOfPlayers())
        Assertions.assertEquals(1, this.players.getNumberOfConnectedPlayers())
    }

    @Test
    fun testJoined() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()

        this.players.joined(
            PlayerJoinedEventDto.newBuilder()
                .setPlayerId(one.getGuid())
                .setShip(5)
                .build()
        )
        val event = eventReference.get()!!.getEvent().getPlayerJoined()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(one.getGuid(), event!!.playerId)
        Assertions.assertEquals(5, event.ship)

        Assertions.assertEquals(1, this.players.getTotalNumberOfPlayers())
    }

    @Test
    fun testLeave() {
        val dtos = ArrayList<Dto>()
        getClientEvents().registerEventDispatcher { e -> dtos.add(e) }

        val one = newId()
        val player = this.players.createOrGet(one)
        this.players.connected(one)
        player.leave()

        val events = dtos
            .filter { it.hasEvent() }
            .map { it.getEvent() }

        Assertions.assertEquals(1, events.size)

        val event = events[0]
        Assertions.assertEquals(one.getGuid(), event.getPlayerLeft().playerId)

        Assertions.assertFalse(
            this.players
                .any { p: Player? -> p === player })

        Assertions.assertEquals(0, this.players.getTotalNumberOfPlayers())
        Assertions.assertEquals(0, this.players.getNumberOfConnectedPlayers())
    }

    @Test
    fun testRequestCurrentView() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        val two = newId()
        this.players.createOrGet(one).join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        this.players.createOrGet(two).join(
            JoinCommandDto
                .newBuilder()
                .setName("two")
                .build()
        )

        this.players.requestCurrentView(one)

        val directedCommandDto = eventReference.get()!!.getDirectedCommand()
        Assertions.assertNotNull(directedCommandDto)

        Assertions.assertTrue(directedCommandDto!!.hasCurrentView())

        val command = directedCommandDto.getCurrentView()

        Assertions.assertNotNull(command)
        Assertions.assertEquals(2, command!!.playersCount)

        val playerIds = command.playersList
            .map { obj -> obj.playerId }
            .map { EntityId.Companion.of(it) }
            .toHashSet()

        Assertions.assertTrue(playerIds.contains(one))
        Assertions.assertTrue(playerIds.contains(two))
    }

    @Test
    fun testMove() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 1025L
        player.move(0.1f, 0.1f, Math.PI.toFloat())
        player.fixedUpdate(1025L)

        Assertions.assertTrue(eventReference.get()!!.hasEvent())
        Assertions.assertTrue(eventReference.get()!!.getEvent().hasPlayerMoved())
        val event = eventReference.get()!!.getEvent().getPlayerMoved()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(0.1f, event!!.x)
        Assertions.assertEquals(0.1f, event.y)
        Assertions.assertEquals(Math.PI.toFloat(), event.angle)
        Assertions.assertEquals((Math.PI / 4f).toFloat(), event.thrustAngle)
    }

    @Test
    fun testLimitedMove() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 1025L
        player.move(0f, 2f, null as MoveReason?)
        player.fixedUpdate(1025L)

        Assertions.assertTrue(eventReference.get()!!.hasEvent())
        Assertions.assertTrue(eventReference.get()!!.getEvent().hasPlayerMoved())
        val event = eventReference.get()!!.getEvent().getPlayerMoved()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(0f, event!!.x, 0.0001f)
        Assertions.assertEquals(1f, event.y)
        Assertions.assertEquals(0f, event.angle)
    }

    @Test
    fun testBounceMove() {
        val eventReference = AtomicReference<Dto>(null)
        getClientEvents().registerEventDispatcher { newValue: Dto? -> eventReference.set(newValue) }

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 1025L

        player.moved(
            PlayerMovedEventDto.newBuilder()
                .setPlayerId(one.getGuid())
                .setAngle(-1f)
                .setThrustAngle(0f)
                .setTimestamp(1025)
                .setVelocity(0f)
                .setX(0.1f)
                .setY(0.1f)
                .build()
        )

        every { clock.getTime() } returns 2025L
        // from (0.1,0.1) should hit (0,0) and bounce to about (0.2,0.2)
        player.move(-0.3f, -0.3f, -1f)
        player.fixedUpdate(2025L)

        Assertions.assertTrue(eventReference.get()!!.hasEvent())
        Assertions.assertTrue(eventReference.get()!!.getEvent().hasPlayerMoved())
        val event = eventReference.get()!!.getEvent().getPlayerMoved()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(0.2f, event!!.x, 0.0001f)
        Assertions.assertEquals(0.2f, event.y, 0.0001f)
        Assertions.assertEquals(-1f, event.angle)
    }

    @Test
    fun testOnlyAngleMove() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 75L
        player.move(0f, 0f, Math.PI.toFloat())
        player.fixedUpdate(75L)

        Assertions.assertTrue(eventReference.get()!!.hasEvent())
        Assertions.assertTrue(eventReference.get()!!.getEvent().hasPlayerMoved())
        val event = eventReference.get()!!.getEvent().getPlayerMoved()

        Assertions.assertNotNull(event)
        Assertions.assertEquals(0f, event!!.x)
        Assertions.assertEquals(0f, event.y)
        Assertions.assertEquals(Math.PI.toFloat(), event.angle)
    }

    @Test
    fun testNoMove() {
        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 50L

        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        player.move(0f, 0f, null as MoveReason?)

        Assertions.assertNull(eventReference.get())
    }

    @Test
    fun testMaxFire() {
        Config.setMaxBolts(2)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()

        player.fire()
        player.fire()
        player.fire()


        Assertions.assertEquals(2, player.getActiveBoltCount())
    }

    @Test
    fun testBoltExhaustion() {
        Config.setMaxBolts(2)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )

        every { clock.getTime() } returns 25L
        player.spawn()

        player.fire()

        Assertions.assertEquals(1, player.getActiveBoltCount())

        player.boltExhausted()

        Assertions.assertEquals(0, player.getActiveBoltCount())
    }

    @Test
    fun testFireDirection() {
        Config.setMaxBolts(2)
        Config.setBoltInertiaEnabled(false)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )

        every { clock.getTime() } returns 25L
        player.spawn()

        every { clock.getTime() } returns 50L
        player.move(0.5f, 0.5f, Math.PI.toFloat())

        val eventReference = AtomicReference<DomainEvent?>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        player.fire()

        var event = eventReference.get()
        Assertions.assertNotNull(event)
        Assertions.assertTrue(event!!.event!!.hasPlayerFired())

        player.fired(event.event.getPlayerFired())

        event = eventReference.get()
        Assertions.assertNotNull(event)
        Assertions.assertTrue(event!!.event!!.hasBoltFired())
        Assertions.assertEquals(Math.PI.toFloat(), event.event.getBoltFired().angle, Float.Companion.MIN_VALUE)
    }

    @Test
    @Disabled
    fun testFireDirectionWithInertia() {
        Config.setMaxBolts(2)
        Config.setBoltInertiaEnabled(true)
        Config.setBoltInertiaProjectionScale(1f)
        Config.setBoltInertiaRejectionScale(1f)
        Config.setBoltInertiaNegativeProjectionScale(1f)
        Config.setPlayerMaxMove(20f)
        Config.setBoltSpeed(20f)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )

        every { clock.getTime() } returns 25L
        player.spawn()

        every { clock.getTime() } returns 1025L
        player.move(0f, 0.5f, 0f)

        val eventReference = AtomicReference<DomainEvent?>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        player.fire()

        var event = eventReference.get()
        Assertions.assertNotNull(event)
        Assertions.assertTrue(event!!.event!!.hasPlayerFired())

        player.fired(event.event.getBoltFired())

        event = eventReference.get()
        Assertions.assertNotNull(event)
        Assertions.assertTrue(event!!.event!!.hasBoltFired())
        Assertions.assertEquals(
            Math.PI.toFloat() / 4f,
            event.event.getBoltFired().angle,
            Float.Companion.MIN_VALUE
        )
    }

    @Test
    fun testCollisionHitBullsEye() {
        Config.setUpdateFrequencyInMillis(1000L)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )

        every { clock.getTime() } returns 25L
        player.spawn()

        every { clock.getTime() } returns 1025L
        player.move(0.5f, 0.5f, null as MoveReason?)

        val positionVector = of(
            1f,
            1f,
            (5 * Math.PI / 4f).toFloat(),
            -0.5f / cos(5 * Math.PI / 4).toFloat(),
            25L,
            PositionVector.DEFAULT_CONFIGURATION
        ).update(1025L)

        Assertions.assertTrue(player.collision(positionVector, 0.05f))
    }

    @Test
    fun testCollisionHitWithinRadius() {
        Config.setUpdateFrequencyInMillis(1000L)
        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 1025L
        player.move(0.54f, 0.54f, null as MoveReason?)
        player.fixedUpdate(1025L)

        val positionVector = of(
            0.45f,
            0.45f,
            (Math.PI / 4).toFloat(),
            0.1f / cos(Math.PI / 4).toFloat(),
            0L,
            PositionVector.DEFAULT_CONFIGURATION
        ).update(1000L)

        Assertions.assertTrue(player.collision(positionVector, 0.05f))
    }

    @Test
    fun testCollisionNoHitInsideSquare() {
        Config.setUpdateFrequencyInMillis(1000L)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 50L
        player.move(0.4f, 0.6f, null as MoveReason?)

        val positionVector = of(
            0.45f,
            0.45f,
            (Math.PI / 4).toFloat(),
            0.1f / cos(Math.PI / 4).toFloat(),
            975L,
            PositionVector.DEFAULT_CONFIGURATION
        ).update(1025L)

        Assertions.assertFalse(player.collision(positionVector, 0.05f))
    }

    @Test
    fun testCollisionNoHitOutsideSquareX() {
        Config.setUpdateFrequencyInMillis(1000L)

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 50L
        player.move(0.2f, 0.2f, null as MoveReason?)

        val positionVector = of(
            0.45f,
            0.45f,
            (Math.PI / 4).toFloat(),
            0.1f / cos(Math.PI / 4).toFloat(),
            25L,
            PositionVector.DEFAULT_CONFIGURATION
        ).update(1025L)

        Assertions.assertFalse(player.collision(positionVector, 0.05f))
    }

    @Test
    fun testCollisionNoHitOutsideSquareY() {
        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()
        every { clock.getTime() } returns 50L
        player.move(0.5f, 0.2f, null as MoveReason?)

        val positionVector = of(
            0.45f,
            0.45f,
            (Math.PI / 4).toFloat(),
            0.1f / cos(Math.PI / 4).toFloat(),
            0L,
            PositionVector.DEFAULT_CONFIGURATION
        ).update(1000L)

        Assertions.assertFalse(player.collision(positionVector, 0.05f))
    }

    @Test
    fun testDestroyedBy() {
        val one = newId()
        val two = newId()
        val player1 = this.players.createOrGet(one)
        val player2 = this.players.createOrGet(two)

        player1.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        player2.join(
            JoinCommandDto
                .newBuilder()
                .setName("two")
                .build()
        )

        every { clock.getTime() } returns 25L
        player1.spawn()
        player2.spawn()

        every { clock.getTime() } returns 50L
        player1.move(0.5f, 0.5f, 1f)

        val domainEvents = ArrayList<DomainEvent>()
        getDomainEvents().registerEventDispatcher { e -> domainEvents.add(e) }

        player1.destroy(two, clock.getTime())

        val events = domainEvents
            .map { it.event }


        Assertions.assertEquals(1, events.size)
        Assertions.assertTrue(events[0]!!.hasPlayerDestroyed())

        verify(exactly = 1) { scoreBoard.updateScore(two, 1) }
    }

    @Test
    fun testSpawn() {
        val eventReference = AtomicReference<Dto?>(null)
        getClientEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        val one = newId()
        val player = this.players.createOrGet(one)
        player.join(
            JoinCommandDto
                .newBuilder()
                .setName("one")
                .build()
        )
        every { clock.getTime() } returns 25L
        player.spawn()

        Assertions.assertTrue(eventReference.get()!!.hasEvent())
        Assertions.assertTrue(eventReference.get()!!.getEvent().hasPlayerSpawned())
        val playerSpawnedEvent = eventReference.get()!!.getEvent().getPlayerSpawned()
        val playerMovedEventDto = playerSpawnedEvent.location
        Assertions.assertEquals(0f, playerMovedEventDto.x)
        Assertions.assertEquals(0f, playerMovedEventDto.y)
        Assertions.assertEquals(0f, playerMovedEventDto.angle)
    }
}